docs/architecture/psa-crypto-implementation-structure.md

```diff
@@ -71,3 +71,105 @@ In case of any error occurring at step 3 or 4, psa_fail_key_creation() is called
+
+Note that this is just a basic guide. In some cases, you won't need to change all the files listed here. In some cases, you may need to change other files.
+
+### PSA standardization
+
+Typically, if there's enough demand for a cryptographic mechanism in Mbed TLS, there's enough demand for it to be part of the official PSA Cryptography specification. Therefore the first step before implementing a new mechanism should be to approach the PSA Cryptography working group in Arm for standardization.
+
+At the time of writing, all cryptographic mechanisms that are accessible through `psa_xxx` APIs in in Mbed TLS are current or upcoming PSA standards. Mbed TLS implements some extensions to the PSA API that offer extra integration customization or extra key policies.
+
+Mbed TLS routinely implements cryptographic mechanisms that are not yet part of a published PSA standard, but that are scheduled to be part of a future version of the standard. The Mbed TLS implementation validates the feasibility of the upcoming PSA standard. The PSA Cryptography working group and the Mbed TLS development team communicate during the elaboration of the new interfaces.
+
+### New functions and macros
+
+If a mechanism requires new functions, they should follow the design guidelines in the PSA Cryptography API specification.
+
+Functions that are part of the current or upcoming API are declared in `include/psa/crypto.h`, apart from structure accessors defined in `include/psa/crypto_struct.h`. Functions that have output buffers have associated sufficient-output-size macros in `include/psa/crypto_sizes.h`.
+
+Constants (algorithm identifiers, key type identifiers, etc.) and associated destructor macros (e.g. `PSA_IS_xxx()`) are defined in `include/psa/crypto_values.h`.
+
+Functions and macros that are not intended for standardization, or that are at a stage where the draft standard might still evolve significantly, are declared in `include/psa/crypto_extra.h`.
+
+The PSA Cryptography API specification defines both names and values for certain kinds of constants: algorithms (`PSA_ALG_xxx`), key types (`PSA_KEY_TYPE_xxx`), ECC curve families (`PSA_ECC_FAMILY_xxx`), DH group families (`PSA_DH_FAMILY_xxx`). If Mbed TLS defines an algorithm or a key type that is not part of a current or upcoming PSA standard, pick a value with the `VENDOR` flag set. If Mbed TLS defines an ECC curve or DH group family that is not part of a current or upcoming PSA standard, define a vendor key type and use the family identifier only with this vendor key type.
+
+New constants must have a test case in `tests/suites/test_suite_psa_crypto_metadata.data` that verifies that `PSA_IS_xxx` macros behave properly with the new constant. New `PSA_IS_xxx` macros must be declared in `tests/suites/test_suite_psa_crypto_metadata.function`.
```

Comment:
This is not a must, as described in https://github.com/Mbed-TLS/mbedtls/pull/6652#discussion_r1047053761.

Reply:
It is! New **constants** need a test case in the `.data` file. New **IS macros** need an IS macro in the `.function` file.

The discussion in #6652 is about a new constant not needing an IS macro.

Reply:
Adding a new test case for a key derivation algorithm to the algorithm classification tests in `.data` file [requires](https://github.com/Mbed-TLS/mbedtls/pull/6652/files#diff-b7b39d85b39c9ea62daa4c7f75334c811dbb66296e4f8f7358617250b5db3693R307) an `_IS_` macro for classification. Please see [this comment.](https://github.com/Mbed-TLS/mbedtls/pull/6652#discussion_r1037175823)

Reply:
In that case, there was a new IS macro in the library, so an IS macro was needed in the .function file. And the values for which the IS macro is true need to have the flag in the value's entry in the .data file.

Reply:
The new IS macro was added just to be able to add a test for the constant to the .data file...

Reply:
In https://github.com/Mbed-TLS/mbedtls/pull/6652, there was no point in adding a new IS macro. An IS macro is only useful if it's more general than `#define IS_FOO(x) ((x) == SOME_CONSTANT)`.

However, since there is a new algorithm constant `PSA_ALG_TLS12_ECJPAKE_TO_PMS`, there should be a corresponding test case in `test_suite_psa_crypto_metadata.data`. This test case does not require any change to the function file. `PSA_ALG_TLS12_ECJPAKE_TO_PMS` does not satisfy any of the [algorithm classification macros](https://github.com/Mbed-TLS/mbedtls/blob/mbedtls-3.3.0/tests/suites/test_suite_psa_crypto_metadata.function#L14) (other than the category check macro `PSA_ALG_IS_DERIVATION` which is handled automatically by the test function). So the entry should be
```
key_derivation_algorithm:PSA_ALG_TLS12_ECJPAKE_TO_PMS:0
```

Reply:
`PSA_ALG_TLS12_ECJPAKE_TO_PMS` isn't the only missing entry in the metadata test suite. I'll make a PR.

Reply:
Here's the PR: https://github.com/Mbed-TLS/mbedtls/pull/6802

Adding a metadata test case for `PSA_ALG_TLS12_ECJPAKE_TO_PMS` causes it to be picked up by automatic test case generation, which reveals that it's incompatible with key agreement.

Reply:
> In #6652, there was no point in adding a new IS macro. An IS macro is only useful if it's more general than `#define IS_FOO(x) ((x) == SOME_CONSTANT)`.
> 
> However, since there is a new algorithm constant `PSA_ALG_TLS12_ECJPAKE_TO_PMS`, there should be a corresponding test case in `test_suite_psa_crypto_metadata.data`. This test case does not require any change to the function file. `PSA_ALG_TLS12_ECJPAKE_TO_PMS` does not satisfy any of the [algorithm classification macros](https://github.com/Mbed-TLS/mbedtls/blob/mbedtls-3.3.0/tests/suites/test_suite_psa_crypto_metadata.function#L14) (other than the category check macro `PSA_ALG_IS_DERIVATION` which is handled automatically by the test function). So the entry should be
> 
> ```
> key_derivation_algorithm:PSA_ALG_TLS12_ECJPAKE_TO_PMS:0
> ```

Ah, I haven't thought of doing such a workaround :) thanks!